packages/engine/src/worker/mod.rs

```diff
@@ -406,31 +465,37 @@ impl WorkerController {
                     self.js.send(Some(sim_id), inbound).await?;
                     pending.active_runner = Language::JavaScript;
                 }
-                Dynamic => return Err(Error::UnexpectedTarget(next.target)),
-                Main => {
+                MessageTarget::Dynamic => return Err(Error::UnexpectedTarget(next.target)),
+                MessageTarget::Main => {
                     log::trace!("Task message came back to main, finishing task");
                     self.finish_task(msg.task_id, sim_id, source, next.payload, msg.shared_store)
-                        .await?
+                        .await?;
                 }
             }
         }
         Ok(())
     }
 
-    /// TODO: DOC
+    /// Adds `source` to the list of currently cancelling languages and if there is a running task
+    /// it sends a [`Cancelled`] message for it to the worker pool.
+    ///
+    /// [`Cancelled`]: TaskResultOrCancelled::Cancelled
```

Comment:
I don't think this is right at all. From reading the method it handles a runner giving confirmation that it has done its bit for cancelling the task. It handles this by:
* Getting the task and checks that the task as a whole has already been told to cancel (where the warn is)
* Checks the task knows that the language (source) that confirmed cancellation has cancelled, otherwise it adds it `!langs.contains(&source)`

And then
Drops the task if the cancellation was from the runner the task was executing in (the active runner), sending a cancellation message (`TaskResultOrCancelled`) to the WorkerPool,

or

It continues by doing nothing:
`// else ignore, since it must be that either the active runner cancelled or completed`